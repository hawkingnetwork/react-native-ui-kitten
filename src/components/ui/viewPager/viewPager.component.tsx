/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */

import React from 'react';
import {
  Animated,
  Easing,
  GestureResponderEvent,
  LayoutChangeEvent,
  PanResponder,
  PanResponderCallbacks,
  PanResponderGestureState,
  PanResponderInstance,
  StyleSheet,
  View,
  ViewProps,
  ViewStyle,
} from 'react-native';
import { I18nLayoutService } from '../support/services';

type ChildElement = React.ReactElement;
type ChildrenProp = ChildElement | ChildElement[];

export interface ViewPagerProps extends ViewProps {
  children: ChildrenProp;
  selectedIndex?: number;
  shouldLoadComponent?: (index: number) => boolean;
  onOffsetChange?: (offset: number) => void;
  onSelect?: (index: number) => void;
}

export type ViewPagerElement = React.ReactElement<ViewPagerProps>;

/**
 * `ViewPager` allows flipping through the "pages".
 *
 * @extends React.Component
 *
 * @property {ReactElement | ReactElement[]} children - Determines children of the component.
 *
 * @property {number} selectedIndex - Determines the index of selected "page".
 *
 * @property {(index: number) => boolean} shouldLoadComponent - Determines loading behavior particular page and can be
 * used for lazy loading.
 *
 * @property {(offset: number) => void} onOffsetChange - Fires on scroll event with current scroll offset.
 *
 * @property {(index: number) => void} onSelect - Fires on "page" select with corresponding index.
 *
 * @property {ViewProps} ...ViewProps - Any props applied to View component.
 *
 * @overview-example ViewPagerSimpleUsage
 *
 * @overview-example ViewPagerLazyLoading
 *
 * @example ViewPagerInlineStyling
 */
export class ViewPager extends React.Component<ViewPagerProps> implements PanResponderCallbacks {

  static defaultProps: Partial<ViewPagerProps> = {
    selectedIndex: 0,
    shouldLoadComponent: (): boolean => true,
  };

  private containerRef: React.RefObject<View> = React.createRef();
  private contentWidth: number = 0;
  private contentOffsetValue: number = 0;
  private contentOffset: Animated.Value = new Animated.Value(this.contentOffsetValue);
  private panResponder: PanResponderInstance = PanResponder.create(this);

  public componentDidMount(): void {
    this.contentOffset.addListener(this.onContentOffsetAnimationStateChanged);
  }

  public componentDidUpdate(prevProps: ViewPagerProps): void {
    if (prevProps.selectedIndex !== this.props.selectedIndex) {
      const index: number = this.props.selectedIndex;
      this.scrollToIndex({ index, animated: true });
    }
  }

  public componentWillUnmount(): void {
    this.contentOffset.removeAllListeners();
  }

  public onMoveShouldSetPanResponder = (event: GestureResponderEvent, state: PanResponderGestureState): boolean => {
    const isHorizontalMove: boolean = Math.abs(state.dx) > 0 && Math.abs(state.dx) > Math.abs(state.dy);

    if (isHorizontalMove) {
      const i18nOffset: number = I18nLayoutService.select(state.dx, -state.dx);
      const nextSelectedIndex: number = this.props.selectedIndex - Math.sign(i18nOffset);

      return nextSelectedIndex >= 0 && nextSelectedIndex < this.getChildCount();
    }

    return false;
  };

  public onPanResponderMove = (event: GestureResponderEvent, state: PanResponderGestureState): void => {
    const i18nOffset: number = I18nLayoutService.select(this.contentWidth, -this.contentWidth);
    const selectedPageOffset: number = this.props.selectedIndex * i18nOffset;

    this.contentOffset.setValue(state.dx - selectedPageOffset);
  };

  public onPanResponderRelease = (event: GestureResponderEvent, state: PanResponderGestureState) => {
    if (Math.abs(state.vx) >= 0.5 || Math.abs(state.dx) >= 0.5 * this.contentWidth) {
      const i18nOffset: number = I18nLayoutService.select(state.dx, -state.dx);
      const index: number = i18nOffset > 0 ? this.props.selectedIndex - 1 : this.props.selectedIndex + 1;
      this.scrollToIndex({ index, animated: true });
    } else {
      const index: number = this.props.selectedIndex;
      this.scrollToIndex({ index, animated: true });
    }
  };

  public scrollToIndex(params: { index: number, animated?: boolean }): void {
    const { index, ...rest } = params;
    const childCount = this.getChildCount() - 1;
    const offset: number = this.contentWidth * (index < 0 ? 0 : index > childCount ? childCount : index);

    this.scrollToOffset({ offset, ...rest });
  }

  public scrollToOffset = (params: { offset: number, animated?: boolean }): void => {
    this.createOffsetAnimation(params).start(this.onContentOffsetAnimationStateEnd);
  };

  private onLayout = (event: LayoutChangeEvent): void => {
    this.contentWidth = event.nativeEvent.layout.width / this.getChildCount();

    this.scrollToIndex({
      index: this.props.selectedIndex,
    });
  };

  private onContentOffsetAnimationStateChanged = (state: { value: number }): void => {
    this.contentOffsetValue = I18nLayoutService.select(-state.value, state.value);

    if (this.props.onOffsetChange) {
      this.props.onOffsetChange(this.contentOffsetValue);
    }
  };

  private onContentOffsetAnimationStateEnd = (result: { finished: boolean }): void => {
    const selectedIndex: number = this.contentOffsetValue / this.contentWidth;

    if (selectedIndex !== this.props.selectedIndex && this.props.onSelect) {
      this.props.onSelect(Math.round(selectedIndex));
    }
  };

  private createOffsetAnimation = (params: { offset: number, animated?: boolean }): Animated.CompositeAnimation => {
    const animationDuration: number = params.animated ? 300 : 0;

    return Animated.timing(this.contentOffset, {
      toValue: I18nLayoutService.select(-params.offset, params.offset),
      easing: Easing.linear,
      duration: animationDuration,
    });
  };

  private renderComponentChild = (source: ChildElement, index: number): ChildElement => {
    const contentView: ChildElement | null = this.props.shouldLoadComponent(index) ? source : null;

    return (
      <View style={styles.contentContainer}>
        {contentView}
      </View>
    );
  };

  private renderComponentChildren = (source: ChildrenProp): ChildElement[] => {
    return React.Children.map(source, this.renderComponentChild);
  };

  private getChildCount = (): number => {
    return React.Children.count(this.props.children);
  };

  private getContainerStyle = (): ViewStyle => {
    return {
      width: `${100 * this.getChildCount()}%`,

      // @ts-ignore: RN has no types for `Animated` styles
      transform: [{ translateX: this.contentOffset }],
    };
  };

  public render(): React.ReactElement<ViewProps> {
    const { style, children, ...restProps } = this.props;

    return (
      <Animated.View
        {...restProps}
        ref={this.containerRef}
        style={[styles.container, style, this.getContainerStyle()]}
        onLayout={this.onLayout}
        {...this.panResponder.panHandlers}>
        {this.renderComponentChildren(children)}
      </Animated.View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
  },
  contentContainer: {
    flex: 1,
    width: '100%',
  },
});
