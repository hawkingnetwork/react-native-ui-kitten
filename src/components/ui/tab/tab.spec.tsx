import React from 'react';
import {
  View,
  ImageProps,
  ImageSourcePropType,
  Image,
  Animated,
} from 'react-native';
import {
  fireEvent,
  render,
  RenderAPI,
  shallow,
} from 'react-native-testing-library';
import { ReactTestInstance } from 'react-test-renderer';
import {
  ApplicationProvider,
  ApplicationProviderProps,
  StyleType,
} from '@kitten/theme';
import {
  Tab,
  TabProps,
} from './tab.component';
import {
  TabBar,
  TabBarProps,
} from './tabBar.component';
import {
  TabView,
  TabViewProps,
} from './tabView.component';
import {
  mapping,
  theme,
} from '../support/tests';
import { ViewPager } from '../viewPager/viewPager.component';

describe('@tab: component checks', () => {

  const Mock = (props?: TabProps): React.ReactElement<ApplicationProviderProps> => {
    return (
      <ApplicationProvider
        mapping={mapping}
        theme={theme}>
        <Tab {...props} />
      </ApplicationProvider>
    );
  };

  it('* empty', () => {
    const component: RenderAPI = render(
      <Mock/>,
    );

    const { output } = shallow(component.getByType(Tab));

    expect(output).toMatchSnapshot();
  });

  it('* title', () => {
    const component: RenderAPI = render(
      <Mock title='title'/>,
    );

    const { output } = shallow(component.getByType(Tab));

    expect(output).toMatchSnapshot();
  });

  it('* icon', () => {
    const iconSource: ImageSourcePropType = { uri: 'https://akveo.github.io/eva-icons/fill/png/128/star.png' };

    const icon = (style: StyleType): React.ReactElement<ImageProps> => {
      return (
        <Image
          style={style}
          source={iconSource}
        />
      );
    };

    const component: RenderAPI = render(
      <Mock icon={icon}/>,
    );

    const { output } = shallow(component.getByType(Tab));

    expect(output).toMatchSnapshot();
  });

  it('* title (styled)', () => {
    const component: RenderAPI = render(
      <Mock
        title='title'
        titleStyle={{
          fontSize: 24,
          color: 'tomato',
        }}
      />,
    );

    const { output } = shallow(component.getByType(Tab));

    expect(output).toMatchSnapshot();
  });

});

describe('@tab-bar: component checks', () => {

  const childTestId0: string = '@tab-bar/child-0';
  const childTestId1: string = '@tab-bar/child-1';

  const Mock = (props?: TabBarProps): React.ReactElement<ApplicationProviderProps> => {
    return (
      <ApplicationProvider
        mapping={mapping}
        theme={theme}>
        <TabBar {...props}>
          {props.children}
        </TabBar>
      </ApplicationProvider>
    );
  };

  const ChildMock = Tab;

  it('* emits onSelect with correct args', () => {
    const onSelect = jest.fn();

    const component = render(
      <Mock onSelect={onSelect}>
        <ChildMock testID={childTestId0}/>
        <ChildMock testID={childTestId1}/>
      </Mock>,
    );

    const child1 = component.getByTestId(childTestId1);

    fireEvent(child1, 'select');

    expect(onSelect).toBeCalledWith(1);
  });

});

describe('@tab-view: component checks', () => {

  const Mock = (props?: TabViewProps): React.ReactElement<ApplicationProviderProps> => {
    return (
      <ApplicationProvider
        mapping={mapping}
        theme={theme}>
        <TabView {...props}>
          {props.children}
        </TabView>
      </ApplicationProvider>
    );
  };

  const ChildMock = (props?: TabProps): React.ReactElement<TabProps> => {
    return (
      <Tab {...props} />
    );
  };

  it('* shouldLoadComponent disables child loading', () => {
    const disabledComponentIndex: number = 1;

    const shouldLoadComponent = jest.fn((...args: any[]) => {
      const index: number = args[0];
      return index !== disabledComponentIndex;
    });

    const component: RenderAPI = render(
      <Mock shouldLoadComponent={shouldLoadComponent}>
        <ChildMock>
          <View/>
        </ChildMock>
        <ChildMock>
          <View/>
        </ChildMock>
      </Mock>,
    );

    const viewPager: ReactTestInstance = component.getByType(ViewPager);
    const containerView: ReactTestInstance = viewPager.findByType(Animated.View);
    const unloadedChild: ReactTestInstance = containerView.props.children[disabledComponentIndex];

    expect(unloadedChild.props.children).toBeNull();
  });
});

