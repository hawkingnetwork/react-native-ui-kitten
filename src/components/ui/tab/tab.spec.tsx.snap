// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@tab: component checks * empty 1`] = `
<Wrapper
  forwardedRef={null}
/>
`;

exports[`@tab: component checks * icon 1`] = `
<Wrapper
  forwardedRef={null}
  icon={[Function]}
/>
`;

exports[`@tab: component checks * title (styled) 1`] = `
<Wrapper
  forwardedRef={null}
  title="title"
  titleStyle={
    Object {
      "color": "tomato",
      "fontSize": 24,
    }
  }
/>
`;

exports[`@tab: component checks * title 1`] = `
<Wrapper
  forwardedRef={null}
  title="title"
/>
`;
